### ISSUE TITLE
Product & Marketing Should Be Able to Send Push Notifications to Users

### ISSUE THREAD

**Issue by Sdaswani** [mozilla-mobile/fenix#373]
### Why/User Benefit/User Problem
Product & Marketing Should Be Able to Send Push Notifications to Users to let them know about all the great features of Fenix.

### What / Requirements
- Fenix has to integrate FCM SDK
- Fenix has to use some concoction of Android Components to be able to register for and receive push notifications

### Acceptance Criteria (how do I know when I’m done?)
Product / Marketing can send a push to Fenix builds
A user can opt out of LP/messaging

**Comment by Sdaswani** [comment]
@bbinto Fennec is actually getting away with not using FCM, but Fenix can't as GCM will be deprecated on April 10 and we can't pin to the old Google Play Services version either. 

https://bugzilla.mozilla.org/show_bug.cgi?id=1510730

**Comment by st3fan** [comment]
This ticket is a duplicate of these two tickets, which are happening to support FCM and ADM:

* https://github.com/mozilla-mobile/android-components/issues/1916
* https://github.com/mozilla-mobile/android-components/issues/1917

**Comment by boek** [comment]
@bbinto to prioritize

**Comment by bbinto** [comment]
I agree with @st3fan as this should be covered by the Leanplum SDK integration. @st3fan would we still have to support/integrate with FCM if we use Leanplum SDK - I feel that's a technical question I can't answer.

**Comment by bifleming** [comment]
Just waiting for Send Tab to be completed.

**Comment by liuche** [comment]
@boek is this ready to be worked on? Follow-up from Backlog grooming yesterday.

**Comment by liuche** [comment]
Send tab is landed, removing AC dep.

**Comment by liuche** [comment]
Probably need to just get keys into LP and then test it.

**Comment by jimbo1qaz** [comment]
https://news.ycombinator.com/item?id=20586299

> A simple change I'd like to see in tech culture is to change how we refer to the notifications feature. We should call them interruptions instead.
> 
> Do I want my meeting interrupted if my spouse calls with an emergency? Absolutely.
> 
> Do I want that meeting interrupted to find out I've been given a $10 Lyft credit? Absolutely not.

I don't want to be interrupted by Firefox push notifications either.

**Comment by sblatz** [comment]
@jimbo1qaz and @lime124 raise good points here. I've filed #4663 to track the ability for users to opt out of these communications. Even with this we should be careful about how frequently we use these. I'd love to touch base with @lwa-moz regarding policy on pushing notifs.

**Comment by lime124** [comment]
@mheubusch is also part of the newly formed mobile marketing communications and may also be thinking about this. :)

**Comment by lnicola** [comment]
> https://vreplay.mozilla.com/replay/showRecordingExternal.html?key=HJI4gtp2yXDG8B3

Is there a non-Flash version available?

**Comment by sblatz** [comment]
Touched base with @lwa-moz. He said that he believes this is working but we still need to go through the full QA process to verify its functionality.

**Comment by liuche** [comment]
@bifleming I don't actually see any code here, so my guess is that this was already in-app, and nothing needed to be done except verification to close this.

**Comment by sblatz** [comment]
@liuche that's correct :)

**Comment by jonalmeida** [comment]
This is dependant on https://github.com/mozilla-mobile/fenix/issues/4063

**Comment by AndiAJ** [comment]
Hi @sblatz , could you please let us know how should we test this? :)

**Comment by AndiAJ** [comment]
Hi, I'm not sure if I took the correct steps to verify this issue, but I **wasn't able to send push notifications from Leanplum  to Fenix 2.0.0-rc.1**
I've tried sending a push notification using the "client_id" but bumped into an error.
Based on the Leanplum events I've added the device to the "Development Devices" and added it as a target

![picturemessage_ttgszwb0 p3f](https://user-images.githubusercontent.com/51314259/64183753-0dd2f200-ce73-11e9-86cc-1d79c92900cb.png)
![Capture](https://user-images.githubusercontent.com/51314259/64183761-0f9cb580-ce73-11e9-992a-cbf1ac6eccfd.PNG)

@lwa-moz Could you please review and help me out? :)

**Comment by lwa-moz** [comment]
@AndiAJ - I tried sending an additional push campaign using just your client ID and it doesn't appear that you've received it.

There's nothing wrong with the targeting or campaign set up as far as I can tell. We'll follow up with Leanplum support to understand why the test messages aren't being received.

**Comment by lwa-moz** [comment]
@AndiAJ 
The issue is stemming from the lack of a notification channel in the message itself. The userId in question is using Android OS 9 which requires the usage of a notification channel.

I ran an API call to verify if there are any channels set. Results below:

{
 "response": [
        {
 "success": true,
 "notificationChannels": []
        }
    ]
}

Currently, there are none set. I would advise looking into using the [addAndroidNotificationChannel](https://docs.leanplum.com/reference#post_api-action-addandroidnotificationchannel) API call.

**Comment by boek** [comment]
@lwa-moz Sorry! we were missing a piece in the Leanplum console that I'm adding today :) I will let you know when we're ready to test

**Comment by AndiAJ** [comment]
@boek - I'll remove the **QA needed** label, please let us know when we can re-test. :)

**Comment by liuche** [comment]
@boek will test today so we don't need to add points for this sprint.

**Comment by sblatz** [comment]
@boek what was found out about this?

**Comment by mheubusch** [comment]
@boek Can you label the channel we discussed:  Firefox news

**Comment by AndiAJ** [comment]
Hi, I've tried to re-test sending push notifications on the latest Nightly Build #12890611 from 16/10 but **without any success.**

I've created [this push notification](https://www.leanplum.com/dashboard#/4828751897952256/messaging/5020814254407680) and identified my device (**Device ID = 09f3c01b-9dd5-44f9-958a-d1aa853afb77**)
![image](https://user-images.githubusercontent.com/51314259/66924947-39312b00-f034-11e9-9e42-f2c7b32ee059.png)

After tapping on "Send Preview" this gets displayed "No push tokens exist for this device"
![image](https://user-images.githubusercontent.com/51314259/66925084-77c6e580-f034-11e9-8bb1-c0b03c7d8494.png)

@sblatz @boek @lwa-moz - Could you please take a look and let me know if I'm doing something wrong?
If that's the case could you please help me out with the actual steps needed to be performed ? :)

**Comment by lwa-moz** [comment]
@AndiAJ after looking at your user profile, you indeed do not have a push token set up. This essentially means that your device does not allow push notifications from the Firefox Preview application. Can you confirm that notifications are allowed for Firefox Preview in your device system settings?

**Comment by AndiAJ** [comment]
@lwa-moz Yes, notifications are allowed.
![rsz_screenshot_20191017-095144](https://user-images.githubusercontent.com/51314259/66986868-59122e80-f0c8-11e9-8b2b-ad961e8e19bb.png)

Could you help me set up the push token ?

**Comment by lwa-moz** [comment]
@boek @AndiAJ 

In the process of troubleshooting for this issue, it appears that **all** Preview users do not have push tokens associated with their devices. They should, as long as notifications are enabled (which is the default state). Both Andi and I can confirm that our devices have push notifications enabled, but no push token associated in the Leanplum dashboard.

Leanplum support suggested: If the push token is not populating, there is more than likely an issue with the integration. [Do] the FCM keys match in both the LP dashboard environment and your FCM dashboard?

**Comment by AndiAJ** [comment]
Thanks @lwa-moz for checking ! :smiley:

I'll remove the QA needed label until the above mentioned matter will be clarified.

**Comment by jonalmeida** [comment]
> Leanplum support suggested: If the push token is not populating, there is more than likely an issue with the integration.

This was my analysis too while investigating https://github.com/mozilla-mobile/android-components/issues/4736

**Comment by liuche** [comment]
@boek will keep an eye on this while AC works on it.

**Comment by liuche** [comment]
We're taking this out bc AC is currently busy with other work.

**Comment by vesta0** [comment]
@liuche can we prioritize the A-C work related to this? I don't see an A-C issue linked.

**Comment by jonalmeida** [comment]
> I don't see an A-C issue linked.

To clarify, there is no A-C work here. We found a crash related to this and reverted the changes, now we're going to be re-landing it as well.

**Comment by liuche** [comment]
@jonalmeida looks like you added this to your sprint, could you summarize what's going on bc I'm confused, as well as what actions Fenix would need to take (if you know)?

**Comment by jonalmeida** [comment]
See https://github.com/mozilla-mobile/android-components/issues/4736#issuecomment-546038672 for details. Leanplum integration was causing all the push things to crash. I'm doing the Fenix things, but there aren't any a-c specifics needed.

**Comment by liuche** [comment]
@boek I think I've lost the state of this bug in my head, could you give me a quick refresh on what is done, or what remains to be done? What is this waiting on?

**Comment by jonalmeida** [comment]
This is going to be my only priority for our next sprint (monday).

**Comment by vesta0** [comment]
Thanks @jonalmeida let's try to land this before code freeze!

**Comment by AndiAJ** [comment]
Hi, I've re-checked this matter on the latest **`Nightly #20490606 from 2/18`** using the following devices, **which I've registered as my Development devices**:
• Google Pixel 3a XL (Android 10)
• Google Pixel 3a (Android 9)
• Huawei Mate 20 Lite (Android 8.1.0)
• Samsung Galaxy S7 (Android 7.0)
• OnePlus A3 (Android 6.0.1)
• LG Nexus 4 (Android 5.1.1)

![image](https://user-images.githubusercontent.com/51314259/74718177-5a9c6080-523a-11ea-8c49-09cbb1d43fcd.png)

I've created [this message](https://www.leanplum.com/dashboard#/4828751897952256/messaging/6114508997001216) and tried to sent it to the above mentioned devices.
![image](https://user-images.githubusercontent.com/51314259/74718481-fcbc4880-523a-11ea-98ca-ab2b4b950979.png)

Success/ Failure of sending the desired Push notification:
| Status | Device & OS| User/Device ID|
| ------------- | ------------- | ------------- |
| ❌  | Google Pixel 3a XL, Android OS 10| 5530ff6e-f0df-4f48-abf3-7027a6d70c16 |
| ✔️  | Google Pixel 3a, Android OS 9 | f255a42c-fba1-4ced-9f03-133cba31da5f |
| ✔️  | Huawei Mate 20 Lite, Android OS 8.1.0 | b7ef0eac-d109-427d-bc85-a89d22ec18c4 |
| ✔️  | Samsung Galaxy S7, Samsung SM-G930F, Android OS 7.0 | 5569dac9-1781-4bb3-b91e-86373836114f |
| ✔️  | OnePlus A3, Android OS 6.0.1 | 81c3486d-6aa0-47a2-83cb-8c34c93401b7 |
| ❌  | LG Nexus 4, Android OS 5.1.1 | f751dc91-db2e-4b55-9e3d-f72ac72c10c2 |

► I've noticed that for the LG Nexus 4 (Android 5.1.1) **no Android push token was generated**, regarding the Google Pixel 3a XL (Android 10) I'm not sure **why the push notification wasn't sent?**

► Another worth mentioning aspect would be regarding **the aspect of the push notification**, namely, I've noticed that **on the Pixel 3a & OnePlus A3 the Fenix Nightly icon isn't displayed**, whereas **on the Huawei Mate 20 Lite, Samsung Galaxy S7 the app icon is properly displayed** in the noification.

![notif](https://user-images.githubusercontent.com/51314259/74719736-45750100-523d-11ea-85d9-33b50ff72ea6.jpg)


►  I've also tried sending the message using a specific User ID (Samsung SM-G930F, Android OS 7.0 - 5569dac9-1781-4bb3-b91e-86373836114f) and noticed that **in fact the message got sent to all my registered test devices**.

I've unregistered all the above mentioned devices and tried again using a specific User ID (Samsung SM-G930F, Android OS 7.0 - 5569dac9-1781-4bb3-b91e-86373836114f) but **still no luck this error message popped up.** 
![image](https://user-images.githubusercontent.com/51314259/74721245-bf0dee80-523f-11ea-8b69-35e7f29f2e64.png)

@jonalmeida & @boek  - Please review the above and advise on how to proceed. ☺️ 
I'll remove the QA needed label until further notice,

**Comment by lwa-moz** [comment]
@AndiAJ - I checked with Leanplum support and they had the following suggestions:

- for LG Nexus 4 (Android 5.1.1): This error usually occurs when push notifications are disabled on the device. If you can confirm push notifications are enabled, which I want to assume is the case, I can ask for more suggested causes.
- Are all of these test cases "previews" of the message, or are you actually sending a push to your test devices?

**Comment by lwa-moz** [comment]
Possibly related to this issue: https://github.com/mozilla-mobile/fenix/issues/8974

The Marketing team is continuing to see issues with Push Tokens being generated as expected on test devices, as was the case in Andi's testing.

@st3fan for visibility.

**Comment by jonalmeida** [comment]
> Possibly related to this issue: #8974

We're waiting for QA to verify the fix for the issue, but if you're also to also test it out on your side, that would be helpful too @lwa-moz . 🙂

**Comment by jonalmeida** [comment]
I'm going to go ahead and close this now since the feature has been tested and follow issues have been filed to address bugs.
